### PR TITLE
[SPARK-31642] Add Pagination Support for Structured Streaming Page

### PR DESCRIPTION
**Pull request by iRakson**
### What changes were proposed in this pull request?
Add Pagination Support for structured streaming page. Now both tables `Active Queries` and `Completed Queries` will have pagination. 
To implement pagination, pagination framework from #7399  is used.
* Also tables will only be shown if there is at least one entry in the table.

### Why are the changes needed?
* This will help users in analysing their structured streaming queries in much better way.
* Other Web UI pages support pagination in their table. So this will make web UI more consistent across pages.
* This can prevent potential OOM errors.

### Does this PR introduce _any_ user-facing change?
Yes. Both tables will support pagination.

### How was this patch tested?

**Before Changes:**
Screenshot 2020-05-28 at 12 01 49 AM

**After Changes:**
Screenshot 2020-05-28 at 12 22 28 AM

### CODE REVIEW

**Comment by iRakson** [comment]
cc @srowen @zsxwing @sarutak @uncleGen Kindly Review

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @gengliangwang as well

**Comment by iRakson** [comment]
retest this please

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by uncleGen** [comment]
Retest this please. Seems like an unrelated failure and will fixed in #28391

**Comment by uncleGen** [comment]
Jenkins retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala [hunk 1]
```diff
@@ -35,11 +40,218 @@ private[ui] class StreamingQueryPage(parent: StreamingQueryTab)
...
+
+  override def pageNumberFormField: String = s"$tableTag.page"
+
+  override def pageLink(page: Int): String = {
+    parameterPath +
+    s"&$pageNumberFormField=$page" +
+    s"&$tableTag.sort=$encodedSortColumn" +
+    s"&$tableTag.desc=$sortDesc" +
+    s"&$pageSizeFormField=$pageSize" +
+    s"#$tableTag"
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala:166]: nit: indent

**Review by uncleGen — COMMENTED** [review]
Bunch of codes seem to be copied from codebase, it might be better to find the way how to reuse them.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/ui/UISeleniumSuite.scala [hunk 2]
```diff
@@ -86,34 +86,36 @@ class UISeleniumSuite extends SparkFunSuite with WebBrowser with Matchers with B
             case _: StreamingQueryException =>
           }
 
-          eventually(timeout(30.seconds), interval(100.milliseconds)) {
+          eventually(timeout(300.seconds), interval(100.milliseconds)) {
```
Comment (iRakson) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/ui/UISeleniumSuite.scala:89]: This was modified accidentally, I will update this.

**Review by srowen — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/ui/UISeleniumSuite.scala [hunk 3]
```diff
@@ -91,29 +91,31 @@ class UISeleniumSuite extends SparkFunSuite with WebBrowser with Matchers with B
             goToUi(spark, "/StreamingQuery")
 
             findAll(cssSelector("h3")).map(_.text).toSeq should contain("Streaming Query")
-            findAll(cssSelector("""#activeQueries-table th""")).map(_.text).toSeq should be {
-              List("Name", "Status", "Id", "Run ID", "Start Time", "Duration", "Avg Input /sec",
-                "Avg Process /sec", "Lastest Batch")
+
+            val arrow = 0x25BE.toChar
```
Comment (sarutak) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/ui/UISeleniumSuite.scala:95]: nit: Why not just simply use a literal like `'\u25BE'` ?

**Comment by uncleGen** [comment]
LGTM. Pending on #28512 accepted and then update this PR.

**Comment by iRakson** [comment]
@srowen @sarutak @uncleGen This PR has been updated.

**Review by sarutak — APPROVED** [review]

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by sarutak** [comment]
I've merged to `master`.
Thanks @iRakson !

**Comment by iRakson** [comment]
Thank You 😊 @sarutak @srowen @uncleGen

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala [hunk 4]
```diff
@@ -35,11 +39,147 @@ private[ui] class StreamingQueryPage(parent: StreamingQueryTab)
...
+  override def headers: Seq[Node] = {
+    val headerAndCss: Seq[(String, Boolean, Option[String])] = {
+      Seq(
+        ("Name", true, None),
+        ("Status", false, None),
+        ("ID", true, None),
+        ("Run ID", true, None),
+        ("Start Time", true, None),
+        ("Duration", true, None),
+        ("Avg Input /sec", true, None),
```
Comment (iRakson) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryPage.scala:160]: @srowen @uncleGen @sarutak Should we add tooltips for these columns.

**Comment by gatorsmile** [comment]
> . I will add snapshots soon.

@iRakson could you update the PR description with the screenshot.

**Comment by iRakson** [comment]
> @iRakson could you update the PR description with the screenshot.

Updated.

**Pull request closed without merging**